Site Branding - Theme colors and logo config

milospp opened the pull request:
**[VIVO GitHub issue](https://github.com/vivo-project/VIVO/issues/4019)**
**[Linked Vitro PR]()**


# What does this pull request do?
This pull request allows administrators to customize institutional branding by defining institutional colors and updating the key color scheme of the current theme.

# What's new?
1. Theme Color Customization
Administrators can now change the following theme colors:
    - Primary Color
    - Secondary Color
    - Accent Color
2. Advanced Color Options
Additional options for detailed customization:
    - Primary Lighter Color
    - Primary Darker Color
    - Link Color
    - Text Color
3. Logo Management
    - Option to upload a custom logo via URL.

# Screenshoots
### Regular Color Settings
<img width="342" alt="image" src="https://github.com/user-attachments/assets/30820c44-4c13-464f-8392-992396e69ce1" />

### Advanced Color Settings
<img width="322" alt="image" src="https://github.com/user-attachments/assets/1b58c8a7-299a-4da3-af35-d76c3c3f1493" />

### Reset Color Option
<img width="225" alt="image" src="https://github.com/user-attachments/assets/0f6cf45f-7e73-464e-8a03-4033c09d9a80" />


# How should this be tested?
## General Testing
## Ensure testing is conducted for every available theme.
### Test 1: Theme Colors
1. Select a theme
2. Customize colors (Primary, Secondary, Accent)
3. Verify that the updated colors are applied across the site
### Test 2: Logo Upload
1. Click Change LOGO
2. Upload a logo image
3. Crop and save the logo
    - Ensure cropping is restricted to a 1:1 or wider aspect ratio.
4. Verify that the new logo is displayed correctly.
### Test 3: Logo Removal
1. Click **Change LOGO**
2. Click **Remove Logo**
3. Confirm that the default logo reappears.

balmas commented:
Some problems noted during testing (Tested on Chrome Version 132.0.6834.160 (Official Build) (arm64))

1. After uploading a logo, then clicking save, it returns to the upload photo page, with the error "The form did not contain a 'datafile' field."  The changes made on the branding page are still applied and the logo is updated, but the only way to leave the page is to click "cancel" 
2. the changes seem to apply automatically most of the time. This is not true when you hit "Reset" color settings, but in most other cases, the changes apply whether or not you click Save
4. If I load a small logo but not a large logo it uses the small logo as the large logo (maybe this is intended I'm not sure)
5. Site Name looks like it's used as the html page title but it doesn't appear anywhere else
6. The copyright text doesn't get updated for any theme (it always just reverts back to the default)
7. The primary color selections are a little confusing. Changing lighter doesn't change the others, but change base does. This makes sense to some degree, but it might be better to list base first.
8. None of the changes seem to impact the Wilma theme, although they do impact the others
9. Link color doesn't seem to impact the tenderfoot theme
10. Generally, there probably needs to be some help text to explain what changing primary color affects, what changing secondary color effects, 
11. When you first load the site information page, both "hide advanced color settings" and "show advanced color settings" links are present

balmas commented:
@milospp  I'm having the same problem testing this as I am with #4039 .  I use docker to test VIVO and I think the problem is that @litvinovg fix for https://github.com/vivo-project/VIVO/issues/3034 is present in the the accompanying Vitro PR for this but not here. I tried cherry-picking the fixes into Vitro, but then ran into other problems with the Docker build. I need a clean build to test.

litvinovg commented on webapp/src/main/webapp/themes/wilma/templates/head.ftl:
```diff
@@ -35,4 +36,19 @@
 
+<style>
+    :root {
+        <#if themePrimaryColorLighter?? && themePrimaryColorLighter != "null">--primary-color-lighter: ${themePrimaryColorLighter};</#if>
```
What does the themePrimaryColorLighter != "null" check should do?

milospp replied:
Yes, _?has_content_ is a much cleaner solution. I replaced this in the code

litvinovg commented on webapp/src/main/webapp/themes/tenderfoot/templates/page/partials/head.ftl:
```diff
@@ -31,4 +31,19 @@
 
+<style>
+    :root {
+        <#if themePrimaryColorLighter?? && themePrimaryColorLighter != "null">--primary-color-lighter: ${themePrimaryColorLighter};</#if>
```
I would use ?has_content to check that variable exists, not null and not empty

balmas commented:
I am able to get this branch working in docker now, but I'll wait for the fixes @litvinovg requested before testing.

balmas commented:
I am still struggling to get this branch working for testing.  Maybe rebasing from main would help.

balmas commented:
@milospp I'm still getting a vitro app rather than a vivo one when I build this branch for running in Docker (using the  installer/example-settings.xml) Is there a different settings file I should be using for the build?

balmas commented:
Ok, starting with a completely clean environment, all my docker images purged, I was finally able to get a vivo war file built out of this branch to load vivo correctly in Docker.  So that's the good news.

I did run into some problems with the functionality. 

(1) In Firefox (138.0.3 (aarch64)) the only way I can get the branding widget to appear is if I switch themes, click on the link to get the little popup that tells me to hit save first to appear, then click save.  Any other sequence fails to do anything:

https://www.awesomescreenshot.com/video/39921799?key=b598235073d45c32311257088db8c04c

There are no errors in the developer console in either browser when the widget fails to display.  The feature is mostly unusable in Firefox box of this.

This does not happen in either Chrome or Safari.

(2) Uploading a logo fails because the URL to which it’s being posted has a hardcoded path element, vivo, and when running from the root, the POST fails with a 400 error (see attached screenshot)

(3) Sometimes changes made in the branding widget don’t seem to take hold.  It’s hard to reproduce this reliably — It seems to happen after multiple edits to the colors.  See video at https://www.awesomescreenshot.com/video/39922335?key=9b44cbc951330b7b5eb94ce311a6eb21 for an example

(4) the branding widget disappears on the Capability Map tab
See video
https://www.awesomescreenshot.com/video/39922383?key=3a2849920f93e6c77ec543c2cc385620

(5) the way the primary colors selector works is still a little hard to understand from the user perspective. If I choose the base color in the middle all three colors change, whereas if I choose the lighter or darker color (top or bottom) it doesn’t affect the others. This makes sense if I know the middle color the base color but it isn’t identified as such in the display. I think it would be better to put the base color on the top and label the three boxes so that the user can understand what is happening.

(6) When switching themes, the prior theme customizations are applied in the display while the theme-specific brand shows up in the branding widget. You have to save the changes in the widget for the new theme’s style to be applied. See video https://www.awesomescreenshot.com/video/39923187?key=67003a1aa1d65aad887ba89212437a4d


(7) there is no way to change the the color of the text and backgrounds in the browse-by navigation menu at document.querySelector("#browse-by > nav") 
[
![Organizations-05-16-2025_05_40_PM](https://github.com/user-attachments/assets/f42b12e7-ffd5-418e-9dc1-0cd2e1fd4d5f)
](screenshot)

(8) the branding widget also disappears on the editing displays. See video
https://www.awesomescreenshot.com/video/39923049?key=dcee246f859179814c4933316483268b

milospp commented:
`brandingStyles` are stored in cache, which needs to be updated whenever the theme changes.
To achieve this, I had to add an if statement inside a generic function, which feels suboptimal.

Does anyone know of a better way to handle updates on theme change?

[/api/src/main/java/edu/cornell/mannlib/vedit/controller/OperationController.java](https://github.com/vivo-project/Vitro/blob/23b2cdb5b7a45d96a6e547e65161359ffa4688a4/api/src/main/java/edu/cornell/mannlib/vedit/controller/OperationController.java#L167-L169)
```
            if (newObj instanceof ApplicationBean) {
                SiteBrandingController.updatethemeBrandingCache(((ApplicationBean) newObj).getThemeDir());
            }
```

balmas commented:
Regarding (1) the firefox issue - I have tried removing my cached data, running in incognito mode, and cannot get this to work no matter what I do. Here's the error I see in the console - it seems to have trouble loading the theme-config.json, I have verified that I can load it manually (e.g. at http://localhost:8080/themes/wilma/theme-config.json) and it clearly is there. 
I tried catching it through debugging statements in the developer console, but it gets lost at line 116 .

<img width="1512" alt="Screenshot 2025-05-21 at 10 29 56 AM" src="https://github.com/user-attachments/assets/89c2a657-ccb9-4356-a970-efcd22aa5e09" />

If others are able to get this to work in Firefox, I guess we could assume it is an odd problem with my environment and move on, but it would be good to have at least a couple of people check that.

All of the other problems appear fixed

Regarding the solution for (6), switching themes, I agree that the solution you implemented isn't ideal. I don't have any other good ideas unfortunately. But I also don't think this is a critical issue, since it doesn't really seem to be a very likely use case anyway. It might be better to just document the behavior than to insert that cache clearing code there. 

I would like at least one other person to verify that the functionality works as it should in Firefox before I approve the PR.

milospp commented:
> Regarding (1) the firefox issue - I have tried removing my cached data, running in incognito mode, and cannot get this to work no matter what I do. Here's the error I see in the console - it seems to have trouble loading the theme-config.json, I have verified that I can load it manually (e.g. at http://localhost:8080/themes/wilma/theme-config.json) and it clearly is there. I tried catching it through debugging statements in the developer console, but it gets lost at line 116 .
> 
> <img alt="Screenshot 2025-05-21 at 10 29 56 AM" width="1512" src="https://private-user-images.githubusercontent.com/644870/445994054-89c2a657-ccb9-4356-a970-efcd22aa5e09.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc5MDg2MTEsIm5iZiI6MTc0NzkwODMxMSwicGF0aCI6Ii82NDQ4NzAvNDQ1OTk0MDU0LTg5YzJhNjU3LWNjYjktNDM1Ni1hOTcwLWVmY2QyMmFhNWUwOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyMlQxMDA1MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iODQ0YjAzZjkwOGJjMzhkN2NiMTZmNmI0YzBmMDVkMTJlZDVhYjI1Y2I4OTQyNGU2N2JlZjgzNzIwOWNlMzkxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.otUuN64JErlSzcARUcTpURhROYzX0yIdSEQcLZbPSxE">
> If others are able to get this to work in Firefox, I guess we could assume it is an odd problem with my environment and move on, but it would be good to have at least a couple of people check that.
> 
> All of the other problems appear fixed
> 
> Regarding the solution for (6), switching themes, I agree that the solution you implemented isn't ideal. I don't have any other good ideas unfortunately. But I also don't think this is a critical issue, since it doesn't really seem to be a very likely use case anyway. It might be better to just document the behavior than to insert that cache clearing code there.
> 
> I would like at least one other person to verify that the functionality works as it should in Firefox before I approve the PR.

I was able to reproduce this bug in Firefox and believe I’ve resolved it. Could you please validate it again? Apologies for the inconvenience, and thank you!

milospp commented:
Fixed

9.1 Fixed alert message


9.2: Added a fallback save mechanism:
Users can now click Submit without first closing the color picker or editor. The changes will still be saved..

> Note (macOS): The native color picker will remain open after saving due to OS behavior.


9.3 Addressed Firefox color picker limitations:
Firefox uses the native OS color picker, which behaves differently across platforms:

On Windows:
<img width="300" alt="image" src="https://github.com/user-attachments/assets/dbf3eeff-63e5-4973-b789-a2886d04e962" />
Mac:
<img width="184" alt="image" src="https://github.com/user-attachments/assets/e23d3b29-bfa6-416a-aa83-a4cb306caf00" />
Linux (I think):
<img width="550" alt="image" src="https://github.com/user-attachments/assets/a1d197c7-84e5-456a-ac7a-eeba7c5119ab" />

Platform behavior summary:
On Windows, the color picker blocks interaction with the browser until it’s closed.
On macOS, users can interact with the browser even while the color picker remains open:
Also I found if a second color picker is opened without closing the first, the original disappears and becomes non-interactive (https://www.awesomescreenshot.com/video/40160102?key=8bc422f983f3704363fb8873bb272caf).

These behaviors cannot be handled with JavaScript, as the native picker opens in a separate OS-level window.
If the user does not close it, I cannot.
Creating a fully custom HTML color picker would be required for full control, but this is likely overkill for the intended functionality. 

---

10. This behavior is intentional and was introduced in my previous update (see point 5).

> This functionality only applies when using the default color settings. If the user manually edits the lighter color, their choice is preserved and won't be overridden unless they explicitly reset it.

Explanation:
- When the user edits the primary (emphasized) base color, both lighter and darker variants are automatically adjusted.
- If the user manually modifies either the lighter or darker color, that specific color becomes "locked" and is no longer updated automatically.
- After saving changes (with a modified primary color), further edits to the base color will no longer affect the lighter or darker variants—unless the user clicks "Reset Color," which removes the lock.

https://www.awesomescreenshot.com/video/40160534?key=3f83734eb8794f9479f7f70b5638be99

If you have any suggestions or concerns about this behavior, feel free to share them!

balmas commented:
Re (10) since we can't distinguish easily between a user actively setting the lighter/darker color and it being automatically set for them by choosing a base color, there isn't a great solution here in IMHO.  Most people will be able to figure out if they click Reset they can start over with the colors so I guess I'm fine with what you have coded.  It could be better but it could also be worse.

Re (9) I think your fix to ensure the changes are saved or canceled as appropriate is also good enough for now. It's definitely not worth writing our own color picker at this point. 

I will add my approval for this PR now (and the accompanying Vitro PR which is really where I should probably have been adding all of these comments..)

milospp commented:
@balmas 

Regarding topic 10, maybe we can explicitly show the state of color input.
For example:
<img width="809" alt="image" src="https://github.com/user-attachments/assets/1b29ba83-b701-4f93-8f80-f74fda2b5810" />

or:
<img width="459" alt="image" src="https://github.com/user-attachments/assets/d8cca53e-16ca-4b73-ae71-76e069e32540" />
<img width="435" alt="image" src="https://github.com/user-attachments/assets/de067c40-1c8c-4024-98a1-be6f89711b36" />

balmas commented:
@milospp yes that’s a good idea. I like the 3rd option you have there the best. With the link displayed vertically between the base and darker color.

milospp commented:
> I think this is working pretty well right now. There are still a few elements that don't get changed with the branding (e.g. the tab background colors and the reset button on the Capability Map page), but frankly it's probably good enough for now.
> 
> I also noticed that in Chrome at least when you upload new logos you have to clear your browser cache for them to display.
> 
> Tested in Chrome, Firefox and Safari.

The Reset button on the Capability Map page is styled in red to represent a "danger" action. This color should always remain red and must not be overridden by primary, secondary, or accent colors. If someone wants to change it, a CSS update will be required.
I’d suggest leaving it as it is, but I’m open to suggestions if you think otherwise.

balmas commented:
I think we should just leave this as-is.  Thank you for the explanations @milospp .  No further requests for changes from me.

litvinovg commented:
I see 404 errors while requesting theme-config.json and cssUploadUtils.js

milospp commented:
Can you check if you have that file in the build target location:
`apache-tomcat-folder/webapps/vivo-main/themes/wilma/theme-config.json`

During build, it should be copied from:
`webapp/src/main/webapp/themes/wilma/theme-config.json`

litvinovg commented:
> Can you check if you have that file in the build target location: `apache-tomcat-folder/webapps/vivo-main/themes/wilma/theme-config.json`
> 
> During build, it should be copied from: `webapp/src/main/webapp/themes/wilma/theme-config.json`

You are right, VIVO part wasn't deployed correctly. Thanks!

litvinovg commented:
@milospp Could you take a look why logo's aren't displayed on any theme but Nemo?
